File: backend/src/modules/app/app.config.ts
export interface AndroidReleaseMetadata {
  platform: 'android';
  latestVersionCode: number;
  latestVersionName: string;
  minimumSupportedVersionCode: number;
  updateRequired: boolean;
  releaseTitle: string;
  releaseNotes: string[];
  apkUrl: string;
  apkSha256: string;
  apkSizeBytes: number;
  publishedAt: string;
}

export const ANDROID_RELEASE_METADATA: AndroidReleaseMetadata = {
  platform: 'android',
  latestVersionCode: 6,
  latestVersionName: '1.0.5',
  minimumSupportedVersionCode: 1,
  updateRequired: false,
  releaseTitle: 'Student OS 1.0.5',
  releaseNotes: [
    'GitHub Releases direct APK installation engine',
    'Remote app configuration and real-time feature flag governance',
    'Automated SHA-256 integrity verification and system package installer support',
    'Polished native typography system with enhanced legibility and vertical rhythm',
  ],
  apkUrl: 'https://github.com/siddgbu-hub/student-os/releases/download/v1.0.5/student-os-v1.0.5.apk',
  apkSha256: '2c551ed52e295458b1e0a9399140c3e47374c31533d5d13647dcd178589fd457',
  apkSizeBytes: 2817377,
  publishedAt: '2026-08-20T17:58:00.000Z',
};

export interface AppFeatureFlags {
  analytics: boolean;
  planner: boolean;
  revision: boolean;
  study: boolean;
  payments: boolean;
  webVersion: boolean;
  newDashboard: boolean;
}

export interface AppAnnouncement {
  id: string;
  title: string;
  message: string;
  actionUrl?: string | null;
  actionText?: string | null;
  dismissible: boolean;
  createdAt: string;
}

export interface RemoteAppConfig {
  minimumSupportedVersion: string;
  minimumSupportedVersionCode: number;
  latestVersion: string;
  latestVersionCode: number;
  recommendedUpdateVersion: string;
  forceUpdate: boolean;
  maintenanceMode: boolean;
  maintenanceMessage: string | null;
  features: AppFeatureFlags;
  webUrl: string;
  githubReleaseUrl: string;
  githubLatestReleaseUrl: string;
  githubLatestApkUrl: string;
  latestApkSha256: string | null;
  helpUrl: string;
  supportEmail: string | null;
  announcements: AppAnnouncement[];
}

export const DEFAULT_REMOTE_APP_CONFIG: RemoteAppConfig = {
  minimumSupportedVersion: '1.0.0',
  minimumSupportedVersionCode: 1,
  latestVersion: '1.0.5',
  latestVersionCode: 6,
  recommendedUpdateVersion: '1.0.5',
  forceUpdate: false,
  maintenanceMode: false,
  maintenanceMessage: null,
  features: {
    analytics: true,
    planner: true,
    revision: true,
    study: true,
    payments: true,
    webVersion: true,
    newDashboard: true,
  },
  webUrl: 'https://studentos.kryvlance.in',
  githubReleaseUrl: 'https://github.com/siddgbu-hub/student-os/releases',
  githubLatestReleaseUrl: 'https://github.com/siddgbu-hub/student-os/releases/tag/v1.0.5',
  githubLatestApkUrl: 'https://github.com/siddgbu-hub/student-os/releases/download/v1.0.5/student-os-v1.0.5.apk',
  latestApkSha256: '2c551ed52e295458b1e0a9399140c3e47374c31533d5d13647dcd178589fd457',
  helpUrl: 'https://studentos.kryvlance.in/help',
  supportEmail: null,
  announcements: [],
};

File: android/app/src/main/java/com/studentos/app/config/AppConfig.kt
package com.studentos.app.config

import android.net.Uri
import kotlinx.serialization.Serializable

const val CANONICAL_WEB_URL = "https://studentos.kryvlance.in"
const val CANONICAL_GITHUB_RELEASE_URL = "https://github.com/siddgbu-hub/student-os/releases"
const val CANONICAL_GITHUB_LATEST_RELEASE_URL = "https://github.com/siddgbu-hub/student-os/releases/tag/v1.0.5"
const val CANONICAL_GITHUB_LATEST_APK_URL = "https://github.com/siddgbu-hub/student-os/releases/download/v1.0.5/student-os-v1.0.5.apk"
const val CANONICAL_HELP_URL = "https://studentos.kryvlance.in/help"

/**
 * Validates that an update/release URL uses HTTPS and points to an authorized GitHub domain.
 */
fun isValidGitHubDownloadUrl(url: String?): Boolean {
    if (url.isNullOrBlank() || !url.startsWith("https://", ignoreCase = true)) {
        return false
    }
    return try {
        val uri = java.net.URI(url)
        val host = uri.host?.lowercase() ?: return false
        host == "github.com" ||
            host.endsWith(".github.com") ||
            host == "objects.githubusercontent.com" ||
            host.endsWith(".githubusercontent.com") ||
            host == "github-releases.githubusercontent.com"
    } catch (_: Exception) {
        false
    }
}

@Serializable
data class AppAnnouncementDto(
    val id: String = "",
    val title: String = "",
    val message: String = "",
    val actionUrl: String? = null,
    val actionText: String? = null,
    val dismissible: Boolean = true,
    val createdAt: String = ""
)

@Serializable
data class RemoteAppConfigDto(
    val minimumSupportedVersion: String = "1.0.0",
    val minimumSupportedVersionCode: Int = 1,
    val latestVersion: String = "1.0.5",
    val latestVersionCode: Int = 6,
    val recommendedUpdateVersion: String = "1.0.5",
    val forceUpdate: Boolean = false,
    val maintenanceMode: Boolean = false,
    val maintenanceMessage: String? = null,
    val features: FeatureFlags = FeatureFlags(),
    val webUrl: String = CANONICAL_WEB_URL,
    val githubReleaseUrl: String = CANONICAL_GITHUB_RELEASE_URL,
    val githubLatestReleaseUrl: String = CANONICAL_GITHUB_LATEST_RELEASE_URL,
    val githubLatestApkUrl: String = CANONICAL_GITHUB_LATEST_APK_URL,
    val latestApkSha256: String? = "2c551ed52e295458b1e0a9399140c3e47374c31533d5d13647dcd178589fd457",
    val helpUrl: String = CANONICAL_HELP_URL,
    val supportEmail: String? = null,
    val announcements: List<AppAnnouncementDto> = emptyList()
)

@Serializable
data class RemoteAppConfigResponse(
    val success: Boolean,
    val data: RemoteAppConfigDto? = null,
    val timestamp: String? = null
)

/**
 * State representing the client update status for GitHub Releases distribution.
 */
sealed class AppUpdateState {
    object UpToDate : AppUpdateState()

    data class OptionalUpdate(
        val latestVersionName: String,
        val latestVersionCode: Int,
        val apkUrl: String,
        val releaseUrl: String,
        val apkSha256: String? = null,
        val releaseNotes: List<String> = emptyList()
    ) : AppUpdateState()

    data class MandatoryUpdate(
        val minimumSupportedVersion: String,
        val minimumSupportedVersionCode: Int,
        val latestVersionName: String,
        val apkUrl: String,
        val releaseUrl: String,
        val apkSha256: String? = null,
        val reason: String? = null
    ) : AppUpdateState()
}

/**
 * Full reactive state consumed by application UI and navigation layers.
 */
data class AppConfigState(
    val isLoading: Boolean = false,
    val isMaintenanceMode: Boolean = false,
    val maintenanceMessage: String? = null,
    val updateState: AppUpdateState = AppUpdateState.UpToDate,
    val featureFlags: FeatureFlags = FeatureFlags(),
    val webUrl: String = CANONICAL_WEB_URL,
    val githubReleaseUrl: String = CANONICAL_GITHUB_RELEASE_URL,
    val githubLatestReleaseUrl: String = CANONICAL_GITHUB_LATEST_RELEASE_URL,
    val githubLatestApkUrl: String = CANONICAL_GITHUB_LATEST_APK_URL,
    val latestApkSha256: String? = null,
    val helpUrl: String = CANONICAL_HELP_URL,
    val supportEmail: String? = null,
    val announcements: List<AppAnnouncementDto> = emptyList(),
    val isUsingCachedConfig: Boolean = false,
    val lastFetchedTimestamp: Long = 0L
)
